src/main/scala/za/co/absa/spark/commons/implicits/StructTypeImplicits.scala

```diff
@@ -0,0 +1,432 @@
+     *
+     * For instance, if given 'a.b', 'a.b.c', 'a.b.c.d' where b and c are arrays the common deepest array
+     * path is 'a.b.c'.
+     *
+     * If any of the arrays are on diverging paths this function returns None.
+     *
+     * The purpose of the function is to determine the order of explosions to be made before the dataframe can be
+     * joined on a field inside an array.
+     *
+     * @param fieldPaths A list of paths to analyze
+     * @return Returns a common array path if there is one and None if any of the arrays are on diverging paths
+     */
+    def getDeepestCommonArrayPath(fieldPaths: Seq[String]): Option[String] = {
+      val arrayPaths = fieldPaths.flatMap(path => getAllArraysInPath(path)).distinct
+
+      if (arrayPaths.nonEmpty && isCommonSubPath(arrayPaths: _*)) {
+        Some(arrayPaths.maxBy(_.length))
+      } else {
+        None
+      }
+    }
+
+    /**
+     * For a field path determines the deepest array path.
+     *
+     * For instance, if given 'a.b.c.d' where b and c are arrays the deepest array is 'a.b.c'.
+     *
+     * @param fieldPath A path to analyze
+     * @return Returns a common array path if there is one and None if any of the arrays are on diverging paths
+     */
+    def getDeepestArrayPath(fieldPath: String): Option[String] = {
+      val arrayPaths = getAllArraysInPath(fieldPath)
+
+      if (arrayPaths.nonEmpty) {
+        Some(arrayPaths.maxBy(_.length))
+      } else {
+        None
+      }
+    }
+
+    /**
+     * Determine the name of a field
+     * Will override to "sourcecolumn" in the Metadata if it exists
+     *
+     * @param field  field to work with
+     * @return       Metadata "sourcecolumn" if it exists or field.name
+     */
+    def getFieldNameOverriddenByMetadata(field: StructField): String = {
+      if (field.metadata.contains(MetadataKeys.SourceColumn)) {
+        field.metadata.getString(MetadataKeys.SourceColumn)
+      } else {
+        field.name
+      }
+    }
+
+    /**
+     * Get paths for all array fields in the schema
+     *
+     * @return Seq of dot separated paths of fields in the schema, which are of type Array
+     */
+    def getAllArrayPaths(): Seq[String] = {
+      schema.fields.flatMap(f => getAllArraySubPaths("", f.name, f.dataType)).toSeq
+    }
+
+    /**
+     * Get a closest unique column name
+     *
+     * @param desiredName A prefix to use for the column name
+     * @param schema      A schema to validate if the column already exists
+     * @return A name that can be used as a unique column name
+     */
+    def getClosestUniqueName(desiredName: String, schema: StructType): String = {
+      var exists = true
+      var columnName = ""
+      var i = 0
+      while (exists) {
+        columnName = if (i == 0) desiredName else s"${desiredName}_$i"
+        exists = schema.fields.exists(_.name.compareToIgnoreCase(columnName) == 0)
+        i += 1
+      }
+      columnName
```

Comment:
If we are against vars, even locals

```scala
    def getClosestUniqueName(desiredName: String, schema: StructType): String = {
      @tailrec
      def getNextName(name: String, index: Int): String = {
        val columnName = if (index == 0) name else s"${name}_$index"
        if (schema.fields.exists(_.name.compareToIgnoreCase(columnName) == 0)) {
          getNextName(name, index)
        } else {
          columnName
        }
      }
      
      getNextName(desiredName, 0)
    }
```

Reply:
Even shorter: 

`def fieldExists(name: String): Boolean = schema.fields.exists(_.name.compareToIgnoreCase(name) == 0)

      if (fieldExists(desiredName)) {
        Iterator.from(1).map(index => s"${desiredName}_${index}")
          .dropWhile(fieldExists).next()
      } else {
        desiredName
      }`

Reply:
⬆️ I like the last solution with the infinite iterator very much!